File: services/api-gateway/src/routes/health.js
const { Router } = require("express");
const pool = require("../db/pool");
const { redis } = require("../cache/redis");
const config = require("../config");

const router = Router();

router.get("/", async (req, res) => {
  let ok = true;

  try {
    await pool.query("SELECT 1");
  } catch {
    ok = false;
  }

  try {
    await redis.ping();
  } catch {
    ok = false;
  }

  try {
    const resp = await fetch(`${config.aggregatorUrl}/health`, {
      signal: AbortSignal.timeout(3000),
    });
    if (!resp.ok) ok = false;
  } catch {
    ok = false;
  }

  res.status(ok ? 200 : 503).json({ status: ok ? "ok" : "degraded" });
});

module.exports = router;

File: services/news-aggregator/routers/health.py
from fastapi import APIRouter
from fastapi.responses import JSONResponse
from db.connection import get_pool
from cache.redis_client import get_redis

router = APIRouter()


@router.get("/health")
async def health():
    ok = True
    try:
        pool = await get_pool()
        await pool.fetchval("SELECT 1")
    except Exception:
        ok = False
    try:
        r = await get_redis()
        await r.ping()
    except Exception:
        ok = False
    return JSONResponse(
        {"status": "ok" if ok else "degraded"},
        status_code=200 if ok else 503,
    )
